index.bs

```diff
@@ -3209,6 +3210,81 @@ opened</dfn> steps given |window|, |type| and |message|.
 
 </div>
 
+## Emulation ## {#module-emulation}
+
+The <dfn export for=modules>emulation</dfn> module contains functionality for
+emulating device-dependent and system-dependent parameters.
+
+### Definition ### {#module-emulation-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+
+EmulationCommand = (
+  emulation.SetViewport
+)
+</pre>
+
+### Types ### {#module-emulation-types}
+
+### Commands ### {#module-emulation-commands}
+
+#### The emulation.setViewport Command ####  {#command-emulation-setViewport}
+
+The <dfn export for=commands>emulation.setViewport</dfn> command emulates specific viewport characteristics (e.g. viewport width and viewport height) on the given top-level browsing context.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+    <pre class="cddl remote-cddl">
+      emulation.SetViewport = {
+        method: "emulation.setViewport",
+        params: emulation.SetViewportParameters
+      };
+
+      emulation.SetViewportParameters = {
+        context: browsingContext.BrowsingContext,
+        ? viewport: emulation.Viewport
+      }
+
+      emulation.Viewport = {
+        width: js-uint,
+        height: js-uint,
+      }
+    </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl">
+     EmptyResult
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for emulation.setViewport">
+
+The [=remote end steps=] with |command parameters| are:
+
+1. Let |context id| be the value of the <code>context</code> field of
+   |command parameters|.
+
+1. Let |context| be the result of [=trying=] to [=get a browsing context=] with |context id|.
+
+1. If |context| is not a [=top-level browsing context=], return [=error=] with [=error code=] [=invalid argument=].
+
+1. If |command parameters| [=map/contains=] "<code>viewport</code>", then:
+    1. Let |viewport parameter| be the <code>viewport</code> field of |command parameters|.
+    1. Let |width| be the <code>width</code> field of |viewport parameter| in CSS pixels.
+    1. Let |height| be the <code>height</code> field of |viewport parameter| in CSS pixels.
+    1. If the |width| or |height| is outside of the implementation-specific or runtime constraints, return an [=error=] with the [=error code=] [=unsupported operation=].
+    1. Change the layout viewport of the |context| to be |width| pixels wide and |height| pixels high and run the [[cssom-view-1#resizing-viewports]] steps.
+1. Else:
+    1. Turn off the viewport emulation on the |context|, reset it to the original value and run the [[cssom-view-1#resizing-viewports]] steps.
```

Comment:
"Turn off" and "reset to original value" aren't really defined. I'm not quite sure how to phrase this, but maybe something like "Set the layout viewport of |context| to the implementation-defined default layout viewport for |context|."

Reply:
Updated